core/thread/inc/ROOT/TReentrantRWLock.hxx

```diff
@@ -0,0 +1,150 @@
+
+   using local_t = LocalCounts*;
+
+   local_t GetLocal() {
+      static thread_local LocalCounts gLocal;
+      return &gLocal;
+   }
+
+   void IncrementReadCount(local_t &local) { ++(local->fReadersCount); }
+
+   template <typename MutexT>
+   void IncrementReadCount(local_t &local, MutexT &) { IncrementReadCount(local); }
+
+   void DecrementReadCount(local_t &local) { --(local->fReadersCount); }
+
+   template <typename MutexT>
+   void DecrementReadCount(local_t &local, MutexT &) { DecrementReadCount(local); }
+
+   bool IsNotCurrentWriter(local_t &local) { return !local->fIsWriter; }
+
+   void SetIsWriter(local_t &local)
+   {
+      // if (fWriteRecurse == std::numeric_limits<decltype(fWriteRecurse)>::max()) {
+      //    ::Fatal("TRWSpinLock::WriteLock", "Too many recursions in TRWSpinLock!");
+      // }
+      ++fWriteRecurse;
+      local->fIsWriter = true;
+   }
+
+   void DecrementWriteCount() { --fWriteRecurse; }
+
+   void ResetIsWriter(local_t &local) { local->fIsWriter = false; }
+
+   size_t GetLocalReadersCount(local_t &local) { return local->fReadersCount; }
+};
+
+struct RecurseCounts {
+   using ReaderColl_t = std::unordered_map<std::thread::id, int>;
+   size_t fWriteRecurse; ///<! Number of re-entry in the lock by the same thread.
+
+   std::thread::id fWriterThread; ///<! Holder of the write lock
+   ReaderColl_t fReadersCount;    ///<! Set of reader thread ids
+
+   using local_t = std::thread::id;
+
+   local_t GetLocal() { return std::this_thread::get_id(); }
+
+   void IncrementReadCount(local_t &local) { ++(fReadersCount[local]); }
+
+   template <typename MutexT>
+   void IncrementReadCount(local_t &local, MutexT &mutex)
+   {
+      std::unique_lock<MutexT> lock(mutex);
+      IncrementReadCount(local);
+   }
+
+   void DecrementReadCount(local_t &local) { --(fReadersCount[local]); }
+
+   template <typename MutexT>
+   void DecrementReadCount(local_t &local, MutexT &mutex)
+   {
+      std::unique_lock<MutexT> lock(mutex);
+      DecrementReadCount(local);
+   }
+
+   bool IsNotCurrentWriter(local_t &local) { return fWriterThread != local; }
+
+   void SetIsWriter(local_t &local)
+   {
+      // if (fWriteRecurse == std::numeric_limits<decltype(fWriteRecurse)>::max()) {
+      //    ::Fatal("TRWSpinLock::WriteLock", "Too many recursions in TRWSpinLock!");
+      // }
+      ++fWriteRecurse;
+      fWriterThread = local;
+   }
+
+   void DecrementWriteCount() { --fWriteRecurse; }
+
+   void ResetIsWriter(local_t & /* local */) { fWriterThread = std::thread::id(); }
+
+   size_t GetLocalReadersCount(local_t &local) { return fReadersCount[local]; }
+
+
+};
+} // Internal
+
+template <typename MutexT = ROOT::TSpinMutex, typename RecurseCountsT = Internal::RecurseCounts>
+class TReentrantRWLock {
+private:
+
+   std::atomic<int> fReaders;           ///<! Number of readers
+   std::atomic<int> fReaderReservation; ///<! A reader wants access
+   std::atomic<int> fWriterReservation; ///<! A writer wants access
```

Comment:
This is similar to the existing TRWSpinLock and is indeed debatable.  The 'advantage' is that in case of 'coding error' (decrementing a zero), the consequence is less dramatic (for if statement using comparison).